File: app/src/main/java/com/bearcave/passageplanning/waypoints/WaypointEditorActivity.java
package com.bearcave.passageplanning.waypoints;

import android.support.design.widget.TextInputEditText;
import android.view.ContextMenu;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.EditText;
import android.widget.TextView;

import com.bearcave.passageplanning.R;
import com.bearcave.passageplanning.base.BaseEditorActivity;
import com.bearcave.passageplanning.waypoints.database.WaypointDAO;
import com.bearcave.passageplanning.thames_tide_provider.web.configurationitems.Gauge;

import butterknife.ButterKnife;

public class WaypointEditorActivity extends BaseEditorActivity<WaypointDAO> {

    private TextInputEditText name;
    private TextInputEditText note;
    private TextInputEditText ukc;
    private TextInputEditText longitude;
    private TextInputEditText latitude;
    private TextInputEditText characteristic;
    private TextView gauge;

    private Integer id = -2;

    @Override
    protected void findViews() {
        super.findViews();
        name =              ButterKnife.findById(this, R.id.name_text);
        note =              ButterKnife.findById(this, R.id.note_text);
        characteristic =    ButterKnife.findById(this, R.id.characteristic_text);
        ukc =               ButterKnife.findById(this, R.id.ukc_text);
        latitude =          ButterKnife.findById(this, R.id.latitude_text);
        longitude =         ButterKnife.findById(this, R.id.longitude_text);
        gauge =             ButterKnife.findById(this, R.id.gauge_name);
        gauge.setText(Gauge.MARGATE.getName());

        registerForContextMenu(gauge);
        gauge.setOnClickListener(this::openContextMenu);
    }

    @Override
    protected void setViewsContent(WaypointDAO waypoint) {
        id = waypoint.getId();
        name.setText(waypoint.getName());
        note.setText(waypoint.getNote());
        characteristic.setText(waypoint.getCharacteristic());
        ukc.setText(String.valueOf(waypoint.getUkc()));
        latitude.setText(waypoint.getLatitudeInSecondFormat());
        longitude.setText(waypoint.getLongitudeInSecondFormat());
        gauge.setText(waypoint.getGauge().getName());
    }

    @Override
    protected int getContentLayoutId() {
        return R.layout.content_waypoint_editor;
    }


    public boolean isAllFilled(){
        if( name.getText().length() == 0
            || ukc.getText().length() == 0
            || latitude.getText().length() == 0
            || longitude.getText().length() == 0)
            return false;
        else return true;
    }

    @Override
    protected WaypointDAO getFilledDAO() {
        return new WaypointDAO(
                id,
                name.getText().toString(),
                note.getText().toString(),
                characteristic.getText().toString(),
                Float.valueOf(ukc.getText().toString()),
                latitude.getText().toString(),
                longitude.getText().toString(),
                Gauge.getByName(gauge.getText().toString())
        );
    }

    @Override
    public void onCreateContextMenu(ContextMenu menu, View v, ContextMenu.ContextMenuInfo menuInfo) {
        super.onCreateContextMenu(menu, v, menuInfo);

        menu.setHeaderTitle(R.string.editor_gauge_chooser_title);

        for( Gauge gauge: Gauge.values()){
            menu.add(Menu.NONE, gauge.getId(), Menu.NONE, gauge.getName());
        }
    }

    @Override
    public boolean onContextItemSelected(MenuItem item) {
        gauge.setText(Gauge.getById(item.getItemId()).getName());
        return super.onContextItemSelected(item);
    }
}

File: app/src/main/java/com/bearcave/passageplanning/waypoints/WaypointEditorActivity.kt
package com.bearcave.passageplanning.waypoints

import android.support.v7.widget.PopupMenu
import android.view.Menu
import android.view.View
import android.widget.TextView
import com.bearcave.passageplanning.R
import com.bearcave.passageplanning.base.BaseEditorActivity
import com.bearcave.passageplanning.tides.utils.Gauge
import com.bearcave.passageplanning.tides.utils.TideCurrent
import com.bearcave.passageplanning.waypoints.database.Waypoint
import com.bearcave.passageplanning.waypoints.position_view.LatitudeFragment
import com.bearcave.passageplanning.waypoints.position_view.LongitudeFragment
import kotlinx.android.synthetic.main.content_waypoint_editor.*

class WaypointEditorActivity : BaseEditorActivity<Waypoint>() {

    private var id: Int = -2

    private var longitude: LongitudeFragment? = null
    private var latitude: LatitudeFragment? = null

    override fun findViews() {
        super.findViews()
        gauge.text = Gauge.SOUTHEND.humanCode
        optionalGauge.text = Gauge.SOUTHEND.humanCode
        tideCurrent.text = TideCurrent.GRAVESEND_REACH.tideCurrentStation

        latitude = supportFragmentManager.findFragmentById(R.id.latitude) as LatitudeFragment?
        longitude = supportFragmentManager.findFragmentById(R.id.longitude) as LongitudeFragment?

        gaugeMenuOpener.setOnClickListener { this.showGaugeMenu(gauge)  }
        optionalGaugeMenuOpener.setOnClickListener { if(optionalGaugeChecker.isChecked) this.showGaugeMenu(optionalGauge)  }
        tideCurrentMenuOpener.setOnClickListener { this.showTideCurrentMenu(it)  }
    }


    override fun setViewsContent(`object`: Waypoint) {
        id = `object`.id
        name.setText(`object`.name)
        note.setText(`object`.note)
        characteristic.setText(`object`.characteristic)
        ukc.setText(`object`.ukc.toString())
        latitude?.position = `object`.latitude
        longitude?.position = `object`.longitude
        gauge.text = `object`.gauge.humanCode
        optionalGauge.text = `object`.optionalGauge.humanCode
        optionalGaugeChecker.isChecked = `object`.gauge.id != `object`.optionalGauge.id
    }

    override val contentLayoutId: Int
        get() = R.layout.content_waypoint_editor


    override val isAllFilled: Boolean
        get() = !(name!!.text.isEmpty()
                    || ukc!!.text.isEmpty())


    override val filledDAO: Waypoint
        get() = Waypoint(
                id,
                name.text.toString(),
                note.text.toString(),
                characteristic.text.toString(),
                ukc.text.toString().toFloat(),
                latitude!!.position,
                longitude!!.position,
                Gauge.getByName(gauge.text.toString()),
                if(optionalGaugeChecker.isChecked)  Gauge.getByName(optionalGauge.text.toString())
                else                                Gauge.getByName(gauge.text.toString()),
                TideCurrent.getByName(tideCurrent!!.text.toString())
        )

    private fun showGaugeMenu(anchor: TextView) {
        val menu = PopupMenu(this, anchor)

        for (gauge in Gauge.values()) {
            menu.menu.add(Menu.NONE, gauge.id, Menu.NONE, gauge.humanCode)
        }

        menu.setOnMenuItemClickListener { item ->
            anchor.text = Gauge.getById(item.itemId).humanCode
            true
        }

        menu.show()
    }

    private fun showTideCurrentMenu(anchor: View){
        val menu = PopupMenu(this, anchor)
        for (tide in TideCurrent.values()) {
            menu.menu.add(Menu.NONE, tide.id, Menu.NONE, tide.tideCurrentStation)
        }

        menu.setOnMenuItemClickListener { item ->
            tideCurrent.text = TideCurrent.getById(item.itemId).tideCurrentStation
            true
        }

        menu.show()
    }
}
